### ISSUE TITLE
Hierarchical clustering with different min-seq-id values

### ISSUE THREAD

**Issue by SimonKitSangChu** [soedinglab/MMseqs2#700]
## Expected Behavior

I wish to have multiple cluster runs, each with a different value of min-seq-id value. Ideally, the resultant representative form hierarchical subsets, i.e. representative sequences for min-seq-id = 50 form a subset of those for min-seq-id = 90, and vice versa.

## Current Behavior
Neither 1) runs of different min-seq-id nor 2) runs of the same min-seq-id agree. 

## Related Issue(s)
https://github.com/soedinglab/MMseqs2/issues/663

**Comment by milot-mirdita** [comment]
You can achieve this by manually cascading your clustering steps. We have implemented something like this a long time ago for Uniclust (https://github.com/milot-mirdita/uniclust-pipeline/blob/9b37347d1e9bb0153efcb986b5d828f9f4a316f0/uniclust_workflow.sh#LL67C7-L128C24). However, this script is not up-to-date, but the general strategy is sound.

Regarding the consistency itself, please refer to this answer:
https://github.com/soedinglab/MMseqs2/issues/115#issuecomment-423111623

**Comment by SimonKitSangChu** [comment]
I am unable to access the uniclust page (page not found). Would you mind sharing another link?

**Comment by milot-mirdita** [comment]
Sorry this one i meant:

https://github.com/soedinglab/uniclust-pipeline/blob/ec52afc312baab5178ce709caebb41bbae9bb3f0/uniclust_workflow.sh#L49-L133

**Comment by milot-mirdita** [comment]
You need to drop all instances of `--max-seq-len` and replace `--target-cov 0.95` with `-c 0.95 --cov-mode 1`.

Lines 112-114 are quite expensive. I think that could be done smarter now, since we have the cluster reassignment mode in the Clustering workflow (which we didn't have back then).

You can replace that with (didn't try it out):
```
mmseqs cluster $INPUT "$TMPPATH/pref_step$STEP" "$TMPPATH/tmp_clu" --cluster-reassign 1 -c 0.8 --min-seq-id 0.3
$RUNNER mmseqs align $INPUT $INPUT "$TMPPATH/pref_step$STEP" "$TMPPATH/aln_step$STEP" -e inf -a
```

**Comment by milot-mirdita** [comment]
This issue is fixed in Git, but not part of a release yet. I think I fixed it shortly after the last release. You can download precompiled binaries at mmseqs.com/latest.

These 5 lines are intended to call the `clust` module which implements the actual clustering algorithm, not the whole `cluster` workflow with the searches etc:
```
mmseqs cluster $INPUT "$TMPPATH/aln_redundancy" "$TMPPATH/clu_redundancy" ${CLUSTER_FRAG_PAR}
mmseqs cluster $INPUT "$TMPPATH/aln_step$STEP" "$TMPPATH/clu_step$STEP" ${CLUSTER0_PAR}
mmseqs cluster $INPUT "$TMPPATH/aln_step$STEP" "$TMPPATH/clu_step$STEP" ${CLUSTER1_PAR}
mmseqs cluster $INPUT "$TMPPATH/aln_uniclust50" "$TMPPATH/clu_uniclust50" ${CLUSTER2_PAR}
mmseqs cluster $INPUT "$TMPPATH/aln_step$STEP" "$TMPPATH/clu_uniclust30" ${CLUSTER2_PAR}
```
should be 

```
mmseqs clust $INPUT "$TMPPATH/aln_redundancy" "$TMPPATH/clu_redundancy" ${CLUSTER_FRAG_PAR}
mmseqs clust $INPUT "$TMPPATH/aln_step$STEP" "$TMPPATH/clu_step$STEP" ${CLUSTER0_PAR}
mmseqs clust $INPUT "$TMPPATH/aln_step$STEP" "$TMPPATH/clu_step$STEP" ${CLUSTER1_PAR}
mmseqs clust $INPUT "$TMPPATH/aln_uniclust50" "$TMPPATH/clu_uniclust50" ${CLUSTER2_PAR}
mmseqs clust $INPUT "$TMPPATH/aln_step$STEP" "$TMPPATH/clu_uniclust30" ${CLUSTER2_PAR}
```

For a 70% clustering you can adapt the 50% clustering strategy:

```
# cluster down to 70%
mmseqs filterdb "$TMPPATH/aln_step$STEP" "$TMPPATH/aln_uniclust70" \
    --filter-column 3 --filter-regex '(0\.[7-9][0-9]{2}|1\.000)'
mmseqs cluster $INPUT "$TMPPATH/aln_uniclust70" "$TMPPATH/clu_uniclust70" ${CLUSTER2_PAR}
mmseqs mergeclusters "${SEQUENCE_DB}" $OUTDIR/uniclust70_$RELEASE \
    "$TMPPATH/clu_frag" "$TMPPATH/clu_redundancy" $TMPPATH/clu_step0 $TMPPATH/clu_step1 $TMPPATH/clu_uniclust70
```

The idea here is to go directly from 90 to 30 to avoid any possible transitivity issues. In the MMseqs2 cascaded clustering you can get representative->member hits below the given thresholds after putting clusters from previous clustering steps together.

You can reconstruct the 50 (and a 70) clustering from the 30% alignments, and that's what it does (it does it with a roundabout regex, because we didn't have expression filters back then).

**Comment by SimonKitSangChu** [comment]
I built mmseqs from the latest Git and the issue is solved. There is now another warning.

```
there must be an error: 37 deleted from 4 that now is empty, but not assigned to a cluster
there must be an error: 49 deleted from 80 that now is empty, but not assigned to a cluster
there must be an error: 42 deleted from 52 that now is empty, but not assigned to a cluster
there must be an error: 42 deleted from 27 that now is empty, but not assigned to a cluster
there must be an error: 128 deleted from 120 that now is empty, but not assigned to a cluster
Total time: 0h 0m 0s 170ms
```

Interestingly it only happens at 50% level. Not 30% or 70%. Let me know if you identify it as a problem.

**Comment by SimonKitSangChu** [comment]
Another issue I notice is that not all representative sequences in 30% are found in 50% or 70%, i.e. the clustering is not "hierarchical".

I added the following lines to extract the representative sequences from the clustered. 
```
mmseqs createsubdb $OUTDIR/uniclust50_${RELEASE} $SEQUENCE_DB $OUTDIR/uniclust50_${RELEASE}_rep
mmseqs convert2fasta $OUTDIR/uniclust50_${RELEASE}_rep uniclust50_rep.fasta
```

**Comment by SimonKitSangChu** [comment]
One more question. Which part of the script should be changed if I am interested in doing the same with foldseek?

**Comment by SimonKitSangChu** [comment]
One more question. Which part of the script should be changed if I am interested in doing the same with foldseek?

**Comment by milot-mirdita** [comment]
Did you solve the issue with the clustering not being hierarchical? This should produce a hierarchical clustering.

Foldseek is quite a bit more tricky since it uses different module will need a lot of parameter changes.

**Comment by SimonKitSangChu** [comment]
Yes, I am still experiencing the same issue, i.e. some sequences in clust30 do not show up in clust50 / clust70.
